File: packages/contracts/python/src/gta_ai_contracts/editorial.py
from __future__ import annotations

from typing import Annotated, Literal
from uuid import UUID

from pydantic import Field, StringConstraints, field_validator

from .common import GameId, StudioModel, UnitScore

ContentType = Literal[
    "vehicle_showcase",
    "vehicle_customization",
    "mission_showcase",
    "mission_guide",
    "tip",
    "secret",
    "myth_test",
    "comparison",
    "challenge",
    "weapon_showcase",
    "location_showcase",
    "activity_showcase",
    "news_explainer",
    "other",
]
TargetPlatform = Literal["tiktok", "youtube_shorts", "youtube_longform"]


class EditorialBrief(StudioModel):
    id: UUID
    project_id: UUID
    schema_version: Literal["1.0"] = "1.0"
    raw_instruction: Annotated[str, StringConstraints(min_length=1)]
    language: Annotated[str, StringConstraints(min_length=2)] = "fr-FR"
    game: GameId = "unknown"
    content_type: ContentType
    subject: str | None = None
    objective: Annotated[str, StringConstraints(min_length=1)]
    target_platforms: list[TargetPlatform]
    target_duration_seconds: Annotated[int, Field(gt=0, le=7_200)] | None = None
    target_aspect_ratio: Literal["9:16", "16:9", "1:1"]
    narrative_order: list[str] = Field(default_factory=list)
    must_include: list[str] = Field(default_factory=list)
    should_include: list[str] = Field(default_factory=list)
    must_avoid: list[str] = Field(default_factory=list)
    expected_events: list[str] = Field(default_factory=list)
    expected_visual_proofs: list[str] = Field(default_factory=list)
    requested_facts: list[str] = Field(default_factory=list)
    requested_comparisons: list[str] = Field(default_factory=list)
    tone: Literal["informative", "enthusiastic", "cinematic", "humorous", "serious", "neutral"]
    pacing: Literal["slow", "balanced", "dynamic", "very_dynamic"]
    spoiler_level: Literal["none", "light", "full"]
    voice_mode: Literal["synthetic_voice", "text_only"]
    publish_mode: Literal["local_export", "approval_required", "automatic"]
    confidence: UnitScore
    ambiguities: list[str] = Field(default_factory=list)

    @field_validator("target_platforms")
    @classmethod
    def unique_platforms(cls, value: list[TargetPlatform]) -> list[TargetPlatform]:
        if not value:
            raise ValueError("at least one target platform is required")
        if len(set(value)) != len(value):
            raise ValueError("target platforms must be unique")
        return value

File: packages/contracts/src/editorial.ts
import type { GameId, Locale, Uuid } from "./common.js";

export const CONTENT_TYPES = [
  "vehicle_showcase",
  "vehicle_customization",
  "mission_showcase",
  "mission_guide",
  "tip",
  "secret",
  "myth_test",
  "comparison",
  "challenge",
  "weapon_showcase",
  "location_showcase",
  "activity_showcase",
  "news_explainer",
  "other",
] as const;

export type ContentType = (typeof CONTENT_TYPES)[number];
export type TargetPlatform = "tiktok" | "youtube_shorts" | "youtube_longform";
export type AspectRatio = "9:16" | "16:9" | "1:1";
export type Tone = "informative" | "enthusiastic" | "cinematic" | "humorous" | "serious" | "neutral";
export type Pacing = "slow" | "balanced" | "dynamic" | "very_dynamic";
export type SpoilerLevel = "none" | "light" | "full";
export type VoiceMode = "synthetic_voice" | "text_only";
export type PublishMode = "local_export" | "approval_required" | "automatic";

export interface EditorialBrief {
  readonly id: Uuid;
  readonly project_id: Uuid;
  readonly schema_version: "1.0";
  readonly raw_instruction: string;
  readonly language: Locale;
  readonly game: GameId;
  readonly content_type: ContentType;
  readonly subject: string | null;
  readonly objective: string;
  readonly target_platforms: readonly TargetPlatform[];
  readonly target_duration_seconds: number | null;
  readonly target_aspect_ratio: AspectRatio;
  readonly narrative_order: readonly string[];
  readonly must_include: readonly string[];
  readonly should_include: readonly string[];
  readonly must_avoid: readonly string[];
  readonly expected_events: readonly string[];
  readonly expected_visual_proofs: readonly string[];
  readonly requested_facts: readonly string[];
  readonly requested_comparisons: readonly string[];
  readonly tone: Tone;
  readonly pacing: Pacing;
  readonly spoiler_level: SpoilerLevel;
  readonly voice_mode: VoiceMode;
  readonly publish_mode: PublishMode;
  readonly confidence: number;
  readonly ambiguities: readonly string[];
}

